### PR TITLE
chore: standardizing packages configs

### PR DESCRIPTION
**Pull request by arboleya**
- Close: #778

### CODE REVIEW

File: packages/abi-coder/package.json [hunk 1]
```diff
@@ -6,16 +6,18 @@
...
     "main": "dist/index.js",
     "module": "dist/index.mjs",
     "types": "dist/index.d.ts",
-    "typings": "dist/index.d.ts",
     "exports": {
       ".": {
         "require": "./dist/index.js",
-        "default": "./dist/index.mjs"
+        "default": "./dist/index.mjs",
+        "types": "./dist/index.d.ts"
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-coder/package.json:20]: Great question.

We have two things here:

 1. The `publishConfig` ([docs](https://pnpm.io/package_json#publishconfig)) is a `pnpm` feature _(not an `npm` one)_ that lets us use an optimal configuration for development and override it with a better alternative for and right before publishing.
 1. The `exports` ([docs](https://nodejs.org/api/packages.html#exports)) is a NodeJS feature that makes it possible to have **_multiple entry-points_**, while **_exporting different versions_** of each one _(mainly `cjs`, `esm`, and `iife`)_ for increased compatibility.

In the example below, we point the `exports` field to our **_Typescript_** file _(optimal for development)_, which enables us to have a full-blown cross-package `go to definition` feature working nicely inside VSCode. Without this, we'd be sent to the `*.d.ts` definitions instead of the source `*.ts` ones, breaking the flow.

```ts
"exports": {
  ".": "./src/index.ts"
,
```

Then we override it for publishing with the following:

```ts
"publishConfig": {
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "default": "./dist/index.mjs",
      "types": "./dist/index.d.ts"
    }
  }
}
```

The fields `main`, `module`, and `types` are a fallback for legacy versions of NodeJS that don't understand the `exports` field. Since the `abi-coder` package exports only a _**single entry point**_ (`index.*`), this is fully equivalent.

However, some packages export multiple entry points _(`abi-typegen`, `fuels`, `versions`, and `wallet`)_, something that is fully supported when using the `exports` syntax but falls short when using the legacy properties approach that allows only a **_single entry point_**.

In those cases, we still have the fallback properties in place, but they'll export only the principal entry point, leaving the others out of the equation. Think of this as a type of _Progressive Enhancement_ approach, targeting newer versions with all functionalities while degrading them to what is possible for legacy ones.

Finally,

You will notice that I removed all legacy fields from the `package.json` ROOT _(in favor of the `exports` syntax)_, leaving them only inside the `publishConfig` container. This is to reduce the confusion between both approaches and **_also_** to straighten up some tiny conflicts I found happening when the two are enabled simultaneously _while pointing to `*.ts` files instead of `*.m?js`_ — and since we don't rely on legacy versions for development, we don't need to care about these _[legacy]_ properties while developing.

File: packages/wallet/src/account.test.ts [hunk 2]
```diff
@@ -13,10 +13,23 @@ import type {
...
+ * from package's index files, using exports syntax such as:
+ *
+ *  export * from '...'
+ *
+ * https://stackoverflow.com/a/72885576
+ */
+jest.mock('@fuel-ts/providers', () => ({
+  __esModule: true,
+  ...jest.requireActual('@fuel-ts/providers'),
+}));
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/wallet/src/account.test.ts:31]: The previous code imported an internal file from the `provider` package **_directly_** (which should not be allowed) to mock it before running the tests. The direct import was only possible because of some _[tiny]_ conflicts between the usage of the legacy `main` field combined with the `exports` one — _when these are pointing to `*.ts` files directly_.

After removing the legacy properties, importing inner modules from packages is not possible anymore. This is to keep things tight and coherent, but the _fix_ bumped into another (ESM / Jest?) issue that required the hack above to keep the mocking step working for these tests.

In short, this workaround _proxies the imported package while injecting the `__esModule: true` property, making it possible for properties to be **redefined** (mocked) in runtime_.

File: packages/abi-coder/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/abi-coder/package.json:20]: Ok, I misspoke, `publishConfig` is indeed an `npm` thing (_as well_):
 - https://docs.npmjs.com/cli/v7/configuring-npm/package-json#publishconfig

**Review by Torres-ssf — APPROVED** [review]
LGTM :clap:

**Review by Dhaiwat10 — APPROVED** [review]
LGTM 🫡
